docs/spec/fastsync/fastsync.md

```diff
@@ -0,0 +1,1163 @@
+- *receivedBlocks*: stores for each height which peer returned
+  it. initially nil
+- *blockstore*: stores for each height greater than
+    *startBlock.Height*, the block of that height. initially nil for
+    all heights
+- *peerTimeStamp*: stores for each peer the last time a block was received
+- *peerRate*: stores for each peer the rate of received data in Bytes/second
+
+#### Auxiliary Functions
+
+#### **[FS-FUNC-TARGET]**:
+- *TargetHeight = max {peerHeigts(addr): addr in peerIDs} union {height}*
+
+#### **[FS-FUNC-MATCH]**:
+
+
+```go
+func VerifyCommit(b Block, c Commit) Boolean
+```
+- Comment
+    - Corresponds to `verifyCommit(chainID string, blockID
+     types.BlockID, height int64, commit *types.Commit) error` in the
+     current Golang implementation, which expects blockID and height 
+	 (from the first block) and the
+     corresponding commit from the following block. We use the
+     simplified form for ease in presentation.
+
+- Implementation remark
+    <!-- - implements the check from -->
+    <!--  [**[TMBC-SOUND-DISTR-PossCommit]**][TMBC-SOUND-DISTR-PossCommit--link], -->
+    <!--  that is, that  *c* is a valid commit for block *b* -->
+	- implements the check that  *c* is a valid commit for block *b*
+- Expected precondition
+    - *c* is a valid commit for block *b*
+- Expected postcondition
+    - *true* if precondition holds
+	- *false* if precondition is violated
+- Error condition
+    - none
+----
+
+## Algorithm Invariants
+
+> In contrast to the temporal properties above that define the problem
+> statement, the following are invariants on the solution to the
+> problem, that is on the algorithm. These invariants are useful for
+> the verification, but can also guide the implementation.
+
+#### **[FS-VAR-STATE-INV]**:
+It is always the case that *state* corresponds to the application state of the
+blockchain of that height, that is, *state = chain[height -
+1].AppState*; *chain* is defined in
+[**[TMBC-SEQ]**][TMBC-SEQ-link].
+
+#### **[FS-VAR-PEER-INV]**:
+It is always the case that the set *peerIDs* only contains nodes that
+have not yet misbehaved (by sending wrong data or timing out).
+
+#### **[FS-VAR-BLOCK-INV]**:
+For *startBlock.Height <= i < height - 1*, let *b(i)* be the block with
+height *i* in *blockstore*, it always holds that
+*VerifyCommit(b(i), b(i+1).Commit) = true*. This means that *height*
+can only be incremented if all blocks with lower height have been verified.
+
+
+## FastSync V2
+
+
+<!--
+> Basic data structures. Simplified, so that we can focus on the distributed
+algorithm here. If existing: link to Tendermint data structures, and mentioned
+if details were omitted. 
+---->
+
+### Outline
+
+<!--
+> Describe solution (in English), decomposition into functions, where communication to other components happens.
+---->
+
+The protocol is described in terms of functions that are triggered by
+(external) events. The implementation uses a scheduler and a
```

Comment:
I am not sure that this sentence is correct. Maybe we just need to explain that some messages are sent periodically (statusRequest) and some actions that lead to sending messages are also triggered periodically. We just need to explain basically execution model.

Reply:
At some point we said we want to keep this underspecified on purpose, now it seems that this and the other comments target at specifying V2 more precisely. We should fix what level of abstraction we are aiming for.